File: level9/source.cpp
class Pown
{
	;
};

Pown *Pown::func (Pown *this, int n)
{
	Pown *ret;

	*this = addr;
	ret = this;
	*(this + 26) = n;
	return ret;
}

void *Pown::setAnnotation (Pown *this, char *s)
{
	return memcpy((char *)this[1], s, strlen(n));
}

int main(int ac, char **av)
{
	Pown *n1, *n2, *n3;

	if (ac < 1)
		exit(1);
	n1 = (Pown *) new(108);
	Pown::func(n1, 5);
	n2 = n1;
	n3 = (Pown *) new(108);
	Pown::func(n3, 6);
	Pown::setAnnotation(n2, av[1]);
	return n3(); // vuln
}

File: level9/source.py
edx = 0
a = {
    init,
    finit
}

def store(a1, a2):
    *a1 = &a
    a1[104] = a2
    return a1

def setAnnotation(a1, str):
    return memcpy(a1 + 4, str, strlen(str)) # We can overflow the stack, and overwrite ret

def main(argc, argv):
    global edx
    if argc <= 1:
        exit(1)
    n1 = new(108)
    store(n1, 5)
    n2 = new(108)
    store(n2, 6)
    setAnnotation(n1, argv[4])
    ret = **n2
    edx = ret
    ret(n2, n1) # Vulnerable call
    return n2

def init(a1, a2):
    return a2[104] + a1[104]

def finit(a1, a2):
    return a1[104] - a2[104]

if __name__ == '__main__':
    import sys
    exit(main(len(sys.argv), sys.argv))
